spherical_splines/SHTOOLS/man/man1/shmultitaperse.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SHMULTITAPERSE 1"
.TH SHMULTITAPERSE 1 "2009-10-19" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "SHMultiTaperSE"
.IX Header "SHMultiTaperSE"
.IP "SHMultiTaperSE \-" 17
.IX Item "SHMultiTaperSE -"
Perform a localized multitaper spectral analysis.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 SHMultiTaperSE (" 28
.IX Item "SUBROUTINE SHMultiTaperSE ("
\&\s-1MTSE\s0, \s-1SD\s0, \s-1SH\s0, \s-1LMAX\s0, \s-1TAPERS\s0, \s-1TAPER_ORDER\s0, \s-1LMAXT\s0, K, \s-1ALPHA\s0, \s-1LAT\s0, \s-1LON\s0, \s-1TAPER_WT\s0, \s-1NORM\s0, \s-1CSPHASE\s0 )
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1MTSE\s0(\s-1LMAX\-LMAXT+1\s0), \s-1SD\s0(\s-1LMAX\-LMAXT+1\s0), \s-1SH\s0(2, \s-1LMAX+1\s0, \s-1LMAX+1\s0), \s-1TAPERS\s0(\s-1LMAXT+1\s0, K)
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1LMAX\s0, \s-1TAPER_ORDER\s0(K), \s-1LMAXT\s0, K
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
\&\s-1ALPHA\s0(3 ), \s-1LAT\s0, \s-1LON\s0, \s-1TAPER_WT\s0(K)
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1NORM\s0, \s-1CSPHASE\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fISHMultiTaperSE\fR will perform a localized multitaper spectral analysis of an input function expressed in spherical harmonics (\s-1SH\s0). The maximum degree of the localized multitaper cross-power spectrum estimate is LMAX-LMAXT. The coefficients and angular orders of the windowing coefficients (\s-1TAPERS\s0 and \s-1TAPER_ORDER\s0) are obtained by a call to \fISHReturnTapers\fR. If \s-1LAT\s0 and \s-1LON\s0 or \s-1ALPHA\s0 is specified, then the symmetry axis of the localizing windows will be rotated to these coordinates. Otherwise, the windows will not be rotated, and the localized spectral analysis will be centered over the north pole.
.PP
If the optional array \s-1TAPER_WT\s0 is specified, then these weights will be used in calculating a weighted average of the individual K tapered estimates (\s-1MTSE\s0) and the corresponding standard error of the estimates (\s-1SD\s0). If not present, the weights will all be assumed to be equal.
.PP
The employed spherical harmonic normalization and Condon-Shortley phase convention can be set by the optional arguments \s-1NORM\s0 and \s-1CSPHASE\s0; if not set, the default is to use geodesy 4\-pi normalized harmonics that exclude the Condon-Shortley phase of (\-1)^m.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1MTSE\s0" 13
.IX Item "MTSE"
(output) REAL*8, \s-1DIMENSION\s0 (\s-1LMAX\-LMAXT+1\s0)
.Sp
The localized multitaper power spectral estimates.
.IP "\s-1SD\s0" 13
.IX Item "SD"
(output) REAL*8, \s-1DIMENSION\s0 (\s-1LMAX\-LMAXT+1\s0)
.Sp
The standard error of the localized multitaper power spectral estimates.
.IP "\s-1SH\s0" 13
.IX Item "SH"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The spherical harmonic coefficients of the function to be localized.
.IP "\s-1LMAX\s0" 13
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The spherical harmonic bandwidth of \s-1SH\s0.
.IP "\s-1TAPERS\s0" 13
.IX Item "TAPERS"
(input) REAL*8, \s-1DIMENSION\s0 (\s-1LMAXT+1\s0, K)
.Sp
An array of the K windowing functions, arranged in columns, obtained from a call to \fISHReturnTapers\fR. Each window has non-zero coefficients for a single angular order that is specified in the array \s-1TAPER_ORDER\s0.
.IP "\s-1TAPER_ORDER\s0" 13
.IX Item "TAPER_ORDER"
(input) \s-1INTEGER\s0, \s-1DIMENSION\s0 (K)
.Sp
An array containing the angular orders of the spherical harmonic coefficients in each column of the array \s-1TAPERS\s0.
.IP "\s-1LMAXT\s0" 13
.IX Item "LMAXT"
(input) \s-1INTEGER\s0
.Sp
The spherical harmonic bandwidth of the windowing functions in the array \s-1TAPERS\s0.
.IP "K" 13
.IX Item "K"
(input) \s-1INTEGER\s0
.Sp
The number of tapers to be utilized in performing the multitaper spectral analysis.
.IP "\s-1ALPHA\s0" 13
.IX Item "ALPHA"
(input) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0(3 )
.Sp
The Euler rotation angles used in rotating the windowing functions. \s-1ALPHA\s0(1 ) = 0, \s-1ALPHA\s0(2 ) = \-(90\-LAT)*PI/180, \s-1ALPHA\s0(3 ) = \-LON*PI/180. Either \s-1ALPHA\s0 or \s-1LAT\s0 and \s-1LON\s0 can be specified, but not both. If none of these are specified, the window functions will not be rotated, and the spectral analysis will be centered at the north pole.
.IP "\s-1LAT\s0" 13
.IX Item "LAT"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The latitude in \s-1DEGREES\s0 of the localized analysis. Either \s-1ALPHA\s0 or \s-1LAT\s0 and \s-1LON\s0 can be specified but not both. If none of these are specified, the window functions will not be rotated, and the spectral analysis will be centered at the north pole.
.IP "\s-1LON\s0" 13
.IX Item "LON"
(input) REAL*8, \s-1OPTIONAL\s0
.Sp
The longitude in \s-1DEGREES\s0 of the localized analysis. Either \s-1ALPHA\s0 or \s-1LAT\s0 and \s-1LON\s0 can be specified, but not both. If none of these are specified, the window functions will not be rotated, and the spectral analysis will be centered at the north pole.
.IP "\s-1TAPER_WT\s0" 13
.IX Item "TAPER_WT"
(input) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0 (K)
.Sp
The weights used in calculating the multitaper spectral estimates and standard error. Optimal values of the weights (for a known global power spectrum) can be obtained from the routine \fISHMTVarOpt\fR.
.IP "\s-1NORM\s0" 13
.IX Item "NORM"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
1 (default) = Geodesy 4\-pi normalized harmonics; 2 = Schmidt semi-normalized harmonics; 3 = unnormalized harmonics; 4 = orthonormal harmonics.
.IP "\s-1CSPHASE\s0" 13
.IX Item "CSPHASE"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
1 (default) = do not apply the Condon-Shortley phase factor to the associated Legendre functions; \-1 = append the Condon-Shortley phase factor of (\-1)^m to the associated Legendre functions.
.SH "NOTES"
.IX Header "NOTES"
If \s-1TAPER_WT\s0 is not specified, the mutltitaper spectral estimate for a given degree will be calculated as the average obtained from the K individual tapered estimates. The standard error of the multitaper estimate at degree l is simply the population standard deviation, S = sqrt( sum (Si \- \s-1MTSE\s0)^2 / (K\-1) ), divided by sqrt(K). If the expectations of each tapered estimate are different, then this method will overestimate the standard error. See Wieczorek and Simons (2007) for the relevant expressions when weigted estimates are used.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIshmultitapercse\fR\|(1), \fIshreturntapers\fR\|(1), \fIshreturntapersm\fR\|(1), \fIshmtvaropt\fR\|(1), \fIshmtvaropt0\fR\|(1), \fIshrotaterealcoef\fR\|(1)
.PP
<http://www.ipgp.jussieu.fr/~wieczor/SHTOOLS/SHTOOLS.html>
.SH "REFERENCES"
.IX Header "REFERENCES"
Wieczorek, M. A. and F. J. Minimum-variance multitaper spectral estimation on the sphere, \fIJ. Fourier Anal. Appl.\fR, 13, doi:10.1007/s00041\-006\-6904\-1, 665\-692, 2007.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2005, 2006 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


spherical_splines/SHTOOLS/src/SHMultiTaperSE.f95
subroutine SHMultiTaperSE(mtse, sd, sh, lmax, tapers, taper_order, lmaxt, K, alpha, lat, lon, &
	taper_wt, norm, csphase)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	This subroutine will calculate the multitaper spectrum estimate utilizing
!	the first K tapers. The standard error is calculated using an unbiased estimate
!	of the sample variance.
!
!	Calling Parameters
!
!		IN
!			sh		Input spherical harmonic file.
!			lmax		Maximum degree of sh.
!			tapers		The eigenvector matrix returned from a program
!					such as SHReturnAllTapers, where each column corresponds
!					to the coefficients for a window for a single non-zero
!					angular order.
!			taper_order	Angular order of the tapers. 
!			lmaxt		Maximum degree of the eigentapers.
!			K		Number of tapers to use in the multitaper spectral estimation.
!		OUT
!			mtse		Multitaper spectrum estimate, valid up to and including a 
!					maximum degree lmax-lmaxt.
!			sd		Standard error of the multitaper spectrum estimate.
!		OPTIONAL (IN)	
!			alpha		Euler angles used to rotate the localizing windows.
!			lat 		Latitude to perform localized analysis (degrees).
!			lon		Longitude to perform localized analysis (degrees).
!			taper_wt	Weight to be applied to each direct spectral estimate. This
!					should sum to unity.
!			csphase:	1: Do not include the phase factor of (-1)^m
!					-1: Apply the phase factor of (-1)^m.
!			norm:		Normalization to be used when calculating Legendre functions
!						(1) "geodesy" (default)
!						(2) Schmidt
!						(3) unnormalized
!						(4) orthonormalized
!
!
!		If the optional parameter alpha (or lat and lon) is input, then the spherical harmonic coefficients of
!		the localizing windows will be rotated accordingly. To rotate a window originally centered
!		at the north pole to (lat, long) given in degrees, use
!
!			alpha(1) = 0.0
!			alpha(2) = -(90.0d0 - lat)*pi/180.0d0
!			alpha(3) =  -long*pi/180.0d0
!
!		See documentation in file ShRotateRealCoef for further information on spherical
!		harmonic rotations.
!
!	Dependencies: SHMultiply, SHPowerSpectrum, SHRotateRealCoef, djpi2, CSPHASE_DEFAULT
!
!	Written by Mark Wieczorek (April 2005)
!		April 2009: Modified routine so that the rotation matrices are only calculated once, instead of
!		for all K tapers.
!
!	Copyright (c) 2009, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use SHTOOLS, only: SHMultiply, SHPowerSpectrum, SHRotateRealCoef, djpi2, CSPHASE_DEFAULT
	
	implicit none
	
	real*8, intent(out) ::	mtse(:), sd(:)
	real*8, intent(in) ::	sh(:,:,:), tapers(:,:)
	integer, intent(in) ::	lmax, lmaxt, K, taper_order(:)
	real*8, intent(in), optional :: alpha(:), lat, lon, taper_wt(:)
	integer, intent(in), optional :: csphase, norm
	integer ::	i, l, phase, mnorm, astat(4)
	real*8 ::	 se(lmax-lmaxt+1, K), x(3), pi, factor
	real*8, allocatable ::	shwin(:,:,:), shloc(:,:,:), dj(:,:,:), shwinrot(:,:,:)
			
	pi = acos(-1.0d0)
	
	if (size(sh(:,1,1)) < 2 .or. size(sh(1,:,1)) < lmax+1 .or. size(sh(1,1,:)) < lmax+1) then
		print*, "ERROR - SHMultiTaperSE"
		print*, "SH must be dimensioned (2,LMAX+1, LMAX+1) where LMAX is ", lmax
		print*, "Input array is dimensioned ", size(sh(:,1,1)), size(sh(1,:,1)), size(sh(1,1,:))
		stop
	elseif(size(tapers(:,1)) < lmaxt+1 .or. size(tapers(1,:)) < K) then
		print*, "ERROR - SHMultiTaperSE"
		print*, "TAPERS must be dimensioned (LMAXT+1, K) where LMAXT and K are, ",lmaxt, K
		print*, "Input array is dimensioned ", size(tapers(:,1)), size(tapers(1,:))
		stop
	elseif(size(taper_order) < K) then
		print*, "ERROR --- SHMutltiTaperSE"
		print*, "TAPER_ORDER must be dimensioned as (K) where K is ", K
		print*, "Input dimension of array is ", size(taper_order)
		stop
	elseif(size(mtse) < lmax-lmaxt+1) then
		print*, "Error --- SHMultiTaperSE"
		print*, "MTSE must be dimensioned as (LMAX-LMAXT+1) where LMAX and LMAXT are ", lmax, lmaxt
		print*, "Input dimension of array is ", size(mtse)
		stop
	elseif(size(sd) < lmax-lmaxt+1) then
		print*, "Error --- SHMultiTaperSE"
		print*, "SD must be dimensioned as (LMAX-LMAXT+1) where LMAX and LMAXT are ", lmax, lmaxt
		print*, "Input dimension of array is ", size(sd)
		stop
	elseif (lmax < lmaxt) then
		print*, "Error --- SHMultiTaperSE"
		print*, "LMAX must be larger than LMAXT."
		print*, "Input valuse of LMAX and LMAXT are ", lmax, lmaxt
		stop
	endif
	
	if(present(taper_wt)) then
		if (size(taper_wt) < K) then
			print*, "ERROR --- SHMultiTaperSE"
			print*, "TAPER_WT must be dimensioned as (K) where K is ", K
			print*, "Input dimension of array is ", size(taper_wt)
			stop
		endif
	endif

	if (present(norm)) then
		if (norm > 4 .or. norm < 1) then
			print*, "Error - SHMultiTaperCSE"
			print*, "Parameter NORM must be 1 (geodesy), 2 (Schmidt), 3 (unnormalized), or 4 (orthonormalized)."
			print*, "Input value is ", norm
			stop
		endif
		mnorm = norm
	else
		mnorm = 1
	endif
	
	if (present(csphase)) then
     		if (csphase /= -1 .and. csphase /= 1) then
     			print*, "Error ---SHMultiTaperSE"
     			print*, "CSPHASE must be 1 (exclude) or -1 (include)."
     			print*, "Input value if ", csphase
     			stop
     		else
     			phase = csphase
     		endif
     	else
     		phase = CSPHASE_DEFAULT
     	endif

	
	if (present(lat) .and. present(lon)) then
		x(1) = 0.0d0
		x(2) = -(90.0d0 - lat)*pi/180.0d0
		x(3) =  -lon*pi/180.0d0
		
	elseif (present(alpha)) then
		if (size(alpha) < 3) then
			print*, "Error --- SHMultiTaperSE"
			print*, "ALPHA must be dimensioned as (3)."
			print*, "Input array is dimensioned as ", size(alpha)
			stop
		endif
		
		x = alpha

	endif

	allocate(shwin(2,lmaxt+1,lmaxt+1), stat = astat(1))
	allocate(shloc(2, lmax+lmaxt+1, lmax+lmaxt+1), stat = astat(2))
	allocate(dj(lmaxt+1,lmaxt+1,lmaxt+1), stat = astat(3))
	allocate(shwinrot(2,lmaxt+1,lmaxt+1), stat = astat(4))
	if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /= 0 .or. astat(4) /= 0) then
		print*, "Error --- SHMultiTaperSE"
		print*, "Problem allocating arrays SHWIN, SHLOC, DJ and SHWINROT", &
			astat(1), astat(2), astat(3), astat(4)
		stop
	endif
	
	shloc = 0.0d0
	mtse = 0.0d0
	sd = 0.0d0
	se = 0.0d0
	dj = 0.0d0
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate localized power spectra
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	if (present(alpha) .or. (present(lat) .and. present(lon)) ) then
		call djpi2(dj, lmaxt)	! Create rotation matrix used in the rotation routine.
	endif

	do i=1, K
		shwin = 0.0d0
		if (taper_order(i) < 0) then
			shwin(2,1:lmaxt+1,abs(taper_order(i))+1) = tapers(1:lmaxt+1,i)
		else
			shwin(1,1:lmaxt+1,taper_order(i)+1) = tapers(1:lmaxt+1,i)
		endif
				
		if (present(alpha) .or. (present(lat) .and. present(lon)) ) then
			call SHRotateRealCoef(shwinrot, shwin, lmaxt, x, dj)
			shwin = shwinrot
		endif
		
		call SHMultiply(shloc, sh, lmax, shwin, lmaxt, csphase = phase, norm = mnorm)
		
		call SHPowerSpectrum(shloc, lmax-lmaxt, se(:,i))
		
	enddo
	
	
	if (present(taper_wt)) then
	
		factor = sum(taper_wt(1:K))**2 - sum(taper_wt(1:K)**2)
		factor = factor * sum(taper_wt(1:K))
		factor= sum(taper_wt(1:K)**2) / factor
	
		do l=0, lmax-lmaxt, 1
			mtse(l+1) = dot_product(se(l+1,1:K), taper_wt(1:K)) / sum(taper_wt(1:K))			
		
			if (K > 1) then 
				sd(l+1) = dot_product( (se(l+1,1:K) - mtse(l+1) )**2, taper_wt(1:K) ) * factor 
			endif
		
		enddo
	
	else
		
		do l=0, lmax-lmaxt, 1
			mtse(l+1) = sum(se(l+1,1:K))/dble(K)
	
			if (K > 1) then 
				sd(l+1) = sum( ( se(l+1,1:K) - mtse(l+1) )**2 ) / dble(K-1) / dble(K) ! standard error!
			endif
		
		enddo
	
	endif
	
	if (K > 1) sd(1:lmax-lmaxt+1) = sqrt(sd(1:lmax-lmaxt+1) )
	
	deallocate(shwin)
	deallocate(shloc)
	deallocate(dj)
	deallocate(shwinrot)

end subroutine SHMultiTaperSE

